### PR TITLE
swarm/network: Refactor peer suggestion engine

### PR DESCRIPTION
**Pull request by nolash**
This PR aims to simplify and cleanly encapsulate the component that informs which peers to attempt connections to, and which peers to request.

The implicit aim is to attain a clear and logical separation of responsbility and functionality between the files `hive.go`, `kademlia.go` and `discovery.go`.

**merge dependencies** (in order of merge to master):

- [x] https://github.com/ethersphere/go-ethereum/pull/1051
- [x] ~~https://github.com/ethersphere/go-ethereum/pull/1074~~
- [x] https://github.com/ethereum/go-ethereum/pull/18408
- [ ] https://github.com/ethersphere/go-ethereum/pull/1069

### CODE REVIEW

File: swarm/network/kademlia.go [hunk 1]
```diff
@@ -463,356 +374,309 @@ func depthForPot(p *pot.Pot, minProxBinSize int, pivotAddr []byte) (depth int) {
...
-			log.Trace(fmt.Sprintf("%08x: ExpNN: %s not found", k.BaseAddr()[:4], pk[:8]))
+			log.Trace(fmt.Sprintf("%08x: ExpNN: %s not found", k.base, pk))
 			culprits = append(culprits, p)
 		}
 	}
-	return gots == len(peers), gots, culprits
+	return connects == len(peers), connects, culprits
+}
+
+// connectedPotential checks whether the peer is connected to a health minimum of peers it knows about in bins that are shallower than depth
```
Comment (zelig) [inline swarm/network/kademlia.go:580]: what is wrong with saturation?

**Review by zelig — COMMENTED** [review]
just took a look. Risky PR

changes a LOT of code around without adding a fix or algorithmic improvement

**Comment by nolash** [comment]
On hold pending success of (along the lines of) https://github.com/ethereum/go-ethereum/pull/18404

**Comment by nolash** [comment]
Discontinued due to https://github.com/ethereum/go-ethereum/pull/18404

**Pull request closed without merging**